### PR TITLE
Service chain errors

### PR DESCRIPTION
**Pull request by ndp-opendap**
The PR improves the HTTP error handling for service chain errors, and returns message regarding the failures in a more understandable manner

### CODE REVIEW

File: dispatch/BESError.h [hunk 1]
```diff
@@ -104,6 +112,16 @@ class BESError: public std::exception,  public BESObj {
...
     }
 
+    /**
+     * Used to add error specific details to the BESInfo object
+     * @param info
+     */
+    virtual void add_my_error_details_to(BESInfo &info) const {
+        // Some errors are covered by the basic details and the message.
+        // Others, like HttpError require more information to be carried
+        // in the error so the override this method.
```
Comment (kyang2014) [inline dispatch/BESError.h:122]: > so the override this method.

Comment (kyang2014) [reply dispatch/BESError.h:122]: Not quite understand the meaning of the above phrase.

File: http/unit-tests/CurlSListTest.cc [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    vector<string> slist_baselines{
+            "FirstName: Willy",
+            "LastName: Wonka",
+            "kjwhebd: jkhbvkwjqehv ljhcljhbwvcqwx",
+            "lorem: "+lorem
+    };
+
+    curl_slist *load_slist(curl_slist *request_headers) const {
+        auto slist = request_headers;
+        //hdrs = curl::append_http_header(hdrs, "Dum", "Dummer");
```
Comment (kyang2014) [inline http/unit-tests/CurlSListTest.cc:92]: Should this commented line be removed? Sonar may issue a code smell.

File: http/unit-tests/CurlSListTest.cc [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            }
+            else {
+                DBG(cerr << prolog << "The slist_itr is nullptr for pass: " << i << "\n");
+            }
+            DBG(cerr << prolog << " slist: " << (void **)slist << " i: " << i << "\n");
+            if((slist != nullptr) && (i < baselines.size())){
+                bool matched = baseline == slist_value;
+                DBG(cerr << prolog << "baseline and slist_value " << (matched?"matched.":"did not match.") << "\n");
+                success = success && matched;
+                // CPPUNIT_ASSERT(matched);
```
Comment (kyang2014) [inline http/unit-tests/CurlSListTest.cc:144]: Another commented line and also some below.

File: http/unit-tests/CurlUtilsTest.cc [hunk 4]
```diff
@@ -148,274 +151,253 @@ class CurlUtilsTest : public CppUnit::TestFixture {
...
+        DBG( cerr << prolog << "END\n");
     }
 
     void add_edl_auth_headers_test() {
-        if (debug) cerr << prolog << "BEGIN" << endl;
-        curl_slist *hdrs = NULL;
-        curl_slist *temp = NULL;
+        DBG( cerr << prolog << "BEGIN\n");
+        curl_slist *hdrs = nullptr;
+        curl_slist *sl_iter;
```
Comment (kyang2014) [inline http/unit-tests/CurlUtilsTest.cc:237]: Why not setting the value  to nullptr

**Review by kyang2014 — APPROVED** [review]
Not familiar with this part of the code.  I just added a few trivial comments. Others look fine to me.

File: http/HttpError.cc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    else {
+        msg << BESIndent::LMarg << "    response_body: [NONE]\n";
+    }
+    BESIndent::UnIndent() ;
+
+    return msg.str();
+}
+
+} // http
```
Comment (jgallagher59701) [inline http/HttpError.cc:60]: Add a new line here at the end of the file

**Review by jgallagher59701 — APPROVED** [review]
I think it would be good to run the auto indent on the .cc and.h files (option+command+L).

The new error stuff looks good; I have no idea about the edits to CurlUtils.cc, but since they fix a ton of errors, they should go in. 

Thanks and sorry for the delay.

File: http/unit-tests/CurlSListTest.cc (hunk 2 above)
Comment (ndp-opendap) [reply http/unit-tests/CurlSListTest.cc]: fixed

File: dispatch/BESError.h (hunk 1 above)
Comment (ndp-opendap) [reply dispatch/BESError.h]: I rewrote it.

**Pull request merged**